src/client.js

```diff
@@ -476,6 +476,113 @@ class ConnectionParameters {
 
 }
 
+// ========================================================================================
+// File Uploader
+// ========================================================================================
+
+const fileUploader = (client) => {
+  /**
+   * Will increament the counter
+   * @returns {Promise<number|*>}
+   */
+  const increaseValue = async () => {
+      const xtkCounter= await client.NLWS.xtkCounter.increaseValue({name: 'xtkResource'});
+      return xtkCounter
+  }
+
+  const createFileRes = (counter, data) => {
+      return {
+          internalName: 'RES' + counter,
+          md5: data[0].md5,
+          label: data[0].fileName,
+          fileName: data[0].fileName,
+          originalName: data[0].fileName,
+          useMd5AsFilename: '1',
+          storageType: 5,
+          xtkschema: 'xtk:fileRes'
+
+      }
+  }
+
+    /**
+     * Will write fileRes
+     * @param fileRes
+     * @returns {Promise<void>}
+     */
+    const write = async (fileRes) => {
+        await client.NLWS.xtkSession.write(fileRes);
+    }
+
+  /**
+   * Will publish the file
+   * @param fileRes
+   * @returns {Promise<void>}
+   */
+  const publishIfNeeded = async (fileRes) => {
+      await client.NLWS.xtkFileRes.create(fileRes).publishIfNeeded();
+  }
+
+  /**
+   * Will return the public URL of the uploaded file.
+   * @param fileRes
+   * @returns {Promise<string|*>}
+   */
+  const getPublicUrl = async (fileRes) => {
+      return await client.NLWS.xtkFileRes.create(fileRes).getURL()
+  }
+  return {
+    upload: async (file) => {
+      return new Promise(async (resolve, reject) => {
+        if (!Util.isBrowser()) {
+          reject('File uploading is only supported in browser based calls.');
+        }
+        try {
+          var data = new FormData()
+          data.append('file_noMd5', file)
+          //TODO: Needs to be refactored after cookie issue get resolved.
+          const response = await fetch(`${client._connectionParameters._endpoint}/nl/jsp/uploadFile.jsp`, {
+            processData: false,
+            credentials: 'include',
+            method: 'POST',
+            body: data,
+            headers: {
+              'x-security-token': client._securityToken,
+              'Cookie': '__sessiontoken=' + client._sessionToken,
+            }
+          })
+          const okay = await response.text()
+          var iframe = document.createElement('iframe');
+          iframe.style.height = 0;
+          iframe.style.width = 0;
+          document.controller = {
+            uploadFileCallBack: async (data) => {
+              const counter = await increaseValue(); // Step 1
+              const fileRes= createFileRes(counter, data)
+              await write(fileRes); // Step 2
+              await publishIfNeeded(fileRes) // Step 3
+              const url = await getPublicUrl(fileRes) // Step 3
+              resolve({
+                label: data[0].fileName,
```

Comment:
Again, it is not clear why data is an array of one element. Maybe more comments would help understand

Reply:
Done